File: js/PaginationHelper.js
// TODO: complete this object/class

// The constructor takes in an array of items and a integer indicating how many
// items fit within a single page
function PaginationHelper(collection, itemsPerPage) {
  this.collection = collection;
  this.pages = collection.reduce((acc, e, i) =>(
    i % itemsPerPage ? acc[acc.length - 1].push(e) : acc.push([e]), acc
  ), []);
}

// returns the number of items within the entire collection
PaginationHelper.prototype.itemCount = function () {
  return this.collection.length;
};

// returns the number of pages
PaginationHelper.prototype.pageCount = function () {
  return this.pages.length;
};

// returns the number of items on the current page. page_index is zero based.
// this method should return -1 for pageIndex values that are out of range
PaginationHelper.prototype.pageItemCount = function (pageIndex) {
  return this.pages[pageIndex] ? this.pages[pageIndex].length : -1;
};

// determines what page an item is on. Zero based indexes
// this method should return -1 for itemIndex values that are out of range
PaginationHelper.prototype.pageIndex = function (itemIndex) {
  const item = this.collection[itemIndex] || null;
  return this.pages.findIndex((page) => page.find((element) => element === item));
};

File: csharp/PaginationHelper.cs
using System;
using System.Collections.Generic;

public class PagnationHelper<T>
{
  // TODO: Complete this class
  
  /// <summary>
  /// Constructor, takes in a list of items and the number of items that fit within a single page
  /// </summary>
  /// <param name="collection">A list of items</param>
  /// <param name="itemsPerPage">The number of items that fit within a single page</param>


  private IList<T> Collection;
  private int ItemsPerPage;
  public PagnationHelper(IList<T> collection, int itemsPerPage)
  {
    Collection = collection;
    ItemsPerPage = itemsPerPage;
  }

  /// <summary>
  /// The number of items within the collection
  /// </summary>
  public int ItemCount
  {
    get
    {
      return Collection.Count;
    }
  }

  /// <summary>
  /// The number of pages
  /// </summary>
  public int PageCount
  {
    get
    {
      return (int)Math.Ceiling((double)ItemCount / (double)ItemsPerPage);
    }
  }

  /// <summary>
  /// Returns the number of items in the page at the given page index 
  /// </summary>
  /// <param name="pageIndex">The zero-based page index to get the number of items for</param>
  /// <returns>The number of items on the specified page or -1 for pageIndex values that are out of range</returns>
  public int PageItemCount(int pageIndex)
  {
    return pageIndex < PageCount && pageIndex >= 0
      ? (int)Math.Min((double)ItemsPerPage, (double)(ItemCount - pageIndex * ItemsPerPage))
      : -1;
  }

  /// <summary>
  /// Returns the page index of the page containing the item at the given item index.
  /// </summary>
  /// <param name="itemIndex">The zero-based index of the item to get the pageIndex for</param>
  /// <returns>The zero-based page index of the page containing the item at the given item index or -1 if the item index is out of range</returns>
  public int PageIndex(int itemIndex)
  {
    return itemIndex < ItemCount && itemIndex >= 0
      ? (int)Math.Floor((double)(itemIndex / ItemsPerPage))
      : -1;
  }
}